126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java

```diff
@@ -0,0 +1,198 @@
+
+@AllArgsConstructor(onConstructor = @__({ @Inject }))
+@Slf4j
+@OwnedBy(HarnessTeam.CDP)
+public class InstanceStatsIteratorRepositoryImpl implements InstanceStatsIteratorRepository {
+  private TimeScaleDBService timeScaleDBService;
+  private CDFeatureFlagHelper cdFeatureFlagHelper;
+  private static final int MAX_RETRY_COUNT = 3;
+
+  public InstanceStatsIterator getLatestRecord(String accountId, String orgId, String projectId, String serviceId)
+      throws Exception {
+    int totalTries = 0;
+    while (totalTries <= MAX_RETRY_COUNT) {
+      ResultSet resultSet = null;
+      try (Connection dbConnection = timeScaleDBService.getDBConnection();
+           PreparedStatement statement = getScopedStatement(dbConnection, accountId, orgId, projectId, serviceId)) {
+        resultSet = statement.executeQuery();
+        return parseInstanceStatsIteratorRecord(resultSet);
+      } catch (SQLException ex) {
+        if (totalTries == MAX_RETRY_COUNT) {
+          log.error("Error while fetching latest instance stats iterator record after all retries");
+          throw ex;
+        }
+        log.warn(
+            "Could not fetch latest instance stats iterator record. Retrying again. Retry number: {}", totalTries, ex);
+        totalTries++;
+      } catch (Exception ex) {
+        log.error("Error while fetching latest instance stats iterator record");
+        throw ex;
+      } finally {
+        DBUtils.close(resultSet);
+      }
+    }
+    return null;
+  }
+
+  public void updateTimestampForIterator(
+      String accountId, String orgId, String projectId, String serviceId, long timestamp) {
+    if (cdFeatureFlagHelper.isEnabled(accountId, FeatureName.CDS_STORE_INSTANCE_STATS_ITERATOR_RUN_TIME)) {
+      boolean successfulOperation = false;
+      int totalTries = 0;
+      while (!successfulOperation && totalTries <= MAX_RETRY_COUNT) {
+        try (Connection dbConnection = timeScaleDBService.getDBConnection();
+             PreparedStatement statement =
+                 getScopedStatementForUpdate(dbConnection, accountId, orgId, projectId, serviceId, timestamp)) {
+          statement.execute();
+          successfulOperation = true;
+        } catch (SQLException ex) {
+          if (totalTries == MAX_RETRY_COUNT) {
+            log.error("Error while updating timestamp for instance stats iterator after all retries");
+          }
+          log.warn("Could not update timestamp for instance stats iterator. Retrying again. Retry number: {}",
+              totalTries, ex);
+          totalTries++;
+        } catch (Exception ex) {
+          log.error("Error while updating timestamp for instance stats iterator", ex);
+          totalTries++;
+        }
+      }
+    }
+  }
+
+  // ------------------------------- PRIVATE METHODS ------------------------------
+  private InstanceStatsIterator parseInstanceStatsIteratorRecord(ResultSet resultSet) throws SQLException {
+    while (resultSet != null && resultSet.next()) {
```

Comment:
```suggestion
    if (resultSet == null || !resultSet.next()) {
       return null;
    } 
    return InstanceStatsIterator.builder()....build();

```
Looks like we aren't looping as such, do we really need while?

Reply:
removed